In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

In [2]:
df = pd.read_parquet("../data/processed/fbref_tm_eda.parquet")

In [3]:
df.head()

,Rk,Player,Nation,Pos,Squad,Comp,Age,Born,Playing Time_MP,Playing Time_Starts,Playing Time_Min,Playing Time_90s,Performance_Gls,Performance_Ast,Performance_G+A,Performance_G-PK,Performance_PK,Performance_PKatt,Performance_CrdY,Performance_CrdR,Per 90 Minutes_Gls,Per 90 Minutes_Ast,Per 90 Minutes_G+A,Per 90 Minutes_G-PK,Per 90 Minutes_G+A-PK,90s,Standard_Gls,Standard_Sh,Standard_SoT,Standard_SoT%,Standard_Sh/90,Standard_SoT/90,Standard_G/Sh,Standard_G/SoT,Standard_PK,Standard_PKatt,Performance_2CrdY,Performance_Fls,Performance_Fld,Performance_Off,Performance_Crs,Performance_Int,Performance_TklW,Performance_OG,Playing Time_Mn/MP,Playing Time_Min%,Starts_Starts,Starts_Mn/Start,Starts_Compl,Subs_Subs,Subs_Mn/Sub,Subs_unSub,Team Success_PPM,Team Success_onG,Team Success_onGA,Team Success_+/-,Team Success_+/-90,Team Success_On-Off,Season,nom,saison_id,player_id,anio_nac,pid_fb,valor_eur,posicion_tm,altura_m,nacimiento,club,contrato_vence,club_coincide,flag_sin_tiros,flag_sin_titularidades,flag_sin_suplencias,flag_cruce_verificado,flag_contrato_conocido,dias_contrato_restante,log_valor_eur,n_metodos_atipico,log1p_Performance_Gls,log1p_Performance_Ast,log1p_Standard_Sh,log1p_Standard_SoT,log1p_Performance_Crs,log1p_Performance_Fls,Age_rs,altura_m_rs,Playing Time_Min_rs,Performance_Gls_rs,Performance_Ast_rs,Standard_Sh_rs,Performance_CrdY_rs,Team Success_PPM_rs
0,1,Max Aarons,ENG,DF,Norwich City,Premier League,21.0,2000.0,34,32,2881,32.0,0,2,2,0,0,0,8,0,0.00,0.06,0.06,0.00,0.06,32.0,0,13,2,15.4,0.41,0.06,0.00,0.00,0,0,0,26,52,1,50,26,45,1,85,84.2,32,89.0,29,2,22.0,2,0.53,18,69,-51,-1.59,0.08,2021-2022,max aarons,2021,471690,2000.0,471690,22000000.0,Right-Back,1.77,2000-01-04,Norwich City,NaT,True,False,False,False,True,False,NaN,7.342423,0,0.000000,1.098612,2.639057,1.098612,3.931826,3.295837,-0.666667,-0.6,1.055258,0.0,1.0,0.20,1.50,-0.949367
1,2,Yunis Abdelhamid,MAR,DF,Reims,Ligue 1,33.0,1987.0,34,34,2983,33.1,2,0,2,2,0,0,5,1,0.06,0.00,0.06,0.06,0.06,33.1,2,18,6,33.3,0.54,0.18,0.11,0.33,0,0,0,38,25,0,4,67,28,0,88,87.2,34,88.0,33,0,0.0,0,1.12,38,41,-3,-0.09,-0.50,2021-2022,yunis abdelhamid,2021,199704,1987.0,199704,1200000.0,Centre-Back,1.90,1987-09-28,Stade Reims,2022-06-30,True,False,False,True,True,True,333.0,6.079181,0,1.098612,0.000000,2.944439,1.945910,1.609438,3.663562,1.333333,0.7,1.115865,1.0,0.0,0.45,0.75,-0.202532
2,3,Salis Abdul Samed,GHA,MF,Clermont Foot,Ligue 1,21.0,2000.0,31,29,2462,27.4,1,0,1,1,0,0,12,3,0.04,0.00,0.04,0.04,0.04,27.4,1,18,5,27.8,0.66,0.18,0.06,0.20,0,0,2,43,37,0,14,40,26,0,79,72.0,29,82.0,19,2,38.0,1,1.10,26,43,-17,-0.62,0.69,2021-2022,salis abdul samed,2021,688707,2000.0,688707,3000000.0,Defensive Midfield,1.79,2000-03-26,Clermont Foot 63,2025-06-30,True,False,False,False,True,True,1429.0,6.477121,0,0.693147,0.000000,2.944439,1.791759,2.708050,3.784190,-0.666667,-0.4,0.806298,0.5,0.0,0.45,2.50,-0.227848
3,4,Laurent Abergel,FRA,MF,Lorient,Ligue 1,28.0,1993.0,34,34,2956,32.8,0,2,2,0,0,0,9,0,0.00,0.06,0.06,0.00,0.06,32.8,0,30,7,23.3,0.91,0.21,0.00,0.00,0,0,0,41,69,1,37,54,63,0,87,86.4,34,87.0,26,0,0.0,0,0.82,27,59,-32,-0.97,-1.75,2021-2022,laurent abergel,2021,238626,1993.0,238626,3000000.0,Defensive Midfield,1.70,1993-02-01,FC Lorient,2025-06-30,True,False,False,True,True,True,1429.0,6.477121,0,0.000000,1.098612,3.433987,2.079442,3.637586,3.737670,0.500000,-1.3,1.099822,0.0,1.0,1.05,1.75,-0.582278
4,5,Charles Abi,FRA,FW,Saint-Étienne,Ligue 1,21.0,2000.0,1,1,45,0.5,0,0,0,0,0,0,0,0,0.00,0.00,0.00,0.00,0.00,0.5,0,0,0,0.0,0.00,0.00,0.00,0.00,0,0,0,1,0,0,1,0,0,0,45,1.3,1,45.0,0,0,0.0,0,1.00,0,0,0,0.00,0.93,2021-2022,charles abi,2021,418658,2000.0,418658,1500000.0,Centre-Forward,1.89,2000-04-12,AS Saint-Étienne,NaT,True,True,False,True,True,False,NaN,6.176091,0,0.000000,0.000000,0.000000,0.000000,0.693147,0.693147,-0.666667,0.6,-0.629828,0.0,0.0,-0.45,-0.50,-0.354430


In [4]:
CHAMPIONS = {
    2021: "Real Madrid",
    2022: "Manchester City",
    2023: "Real Madrid",
    2024: "PSG",
    2025: "PSG",
}

LIGAS = {
    ("Premier League", 2021): "Manchester City",
    ("Premier League", 2022): "Manchester City",
    ("Premier League", 2023): "Manchester City",
    ("Premier League", 2024): "Liverpool",
    ("Premier League", 2025): "Arsenal",
    ("La Liga", 2021): "Real Madrid",
    ("La Liga", 2022): "Barcelona",
    ("La Liga", 2023): "Real Madrid",
    ("La Liga", 2024): "Barcelona",
    ("La Liga", 2025): "Barcelona",
    ("Bundesliga", 2021): "Bayern Munich",
    ("Bundesliga", 2022): "Bayern Munich",
    ("Bundesliga", 2023): "Leverkusen",
    ("Bundesliga", 2024): "Bayern Munich",
    ("Bundesliga", 2025): "Bayern Munich",
    ("Serie A", 2021): "Milan",
    ("Serie A", 2022): "Napoli",
    ("Serie A", 2023): "Inter",
    ("Serie A", 2024): "Napoli",
    ("Serie A", 2025): "Inter",
    ("Ligue 1", 2021): "PSG",
    ("Ligue 1", 2022): "PSG",
    ("Ligue 1", 2023): "PSG",
    ("Ligue 1", 2024): "PSG",
    ("Ligue 1", 2025): "PSG",
}

df['Ganador_Champions'] = [int(CHAMPIONS.get(t) == s) for t, s in zip(df['saison_id'], df['Squad'])]
df['Ganador_Liga'] = [int(LIGAS.get((c, t)) == s)for c, t, s in zip(df['Comp'], df['saison_id'], df['Squad'])]

# Ingeniería de variables para el modelado de `valor_eur`

Partimos de `data/processed/fbref_tm_eda.parquet`, el dataset ya tratado en el EDA (`02-eda.ipynb`): sin duplicados de tipo, con nulos estructurales imputados (y sus flags), atípicos conservados (son la señal de mercado) y con columnas auxiliares ya normalizadas (`_rs`) o log-transformadas (`log1p_`).

**Objetivo de este notebook:** dejar `df` listo como matriz de variables para predecir `valor_eur` (o `log_valor_eur`, su versión log-transformada). Para ello:

1. Se corrigen errores de cruce FBref↔Transfermarkt (filas donde las estadísticas de un jugador quedaron pegadas al valor de mercado de otro).
2. Se eliminan identificadores, columnas de auditoría de calidad de datos y una variable con **fuga de información**.
3. Se eliminan duplicados exactos y variables derivadas (combinaciones lineales de otras que ya se conservan).
4. Se resuelve la redundancia entre columnas "en bruto" y sus versiones ya tratadas en el EDA (`_rs` robust-scaled, `log1p_`), quedándonos con una sola representación por variable.
5. Se construyen **variables categóricas (dummies)** para posición, competición y equipo.
6. Se documenta, variable por variable, el tratamiento aplicado y el motivo.

Cada sección de código va acompañada de una celda markdown que explica **qué se hace y por qué**, para que el tratamiento quede trazable.

## 1. Diagnóstico previo: qué trae `df` heredado del EDA

Antes de decidir qué eliminar o transformar, revisamos tipos de dato, nulos restantes y cardinalidad de las categóricas, y detectamos columnas duplicadas o derivadas por combinación lineal de otras (evidencian redundancia exacta, no solo correlación alta).

In [5]:
print(f"Shape inicial: {df.shape}")
print()
print("Nulos restantes:")
print(df.isna().sum().pipe(lambda s: s[s > 0]))
print()
print("Cardinalidad de columnas categóricas / booleanas:")
for c in df.columns:
    if df[c].dtype.name in ("category", "object", "bool"):
        print(f"  {c:26s} {df[c].nunique():>4d} categorías")

Shape inicial: (13835, 95)

Nulos restantes:
anio_nac                   686
pid_fb                     686
nacimiento                   2
contrato_vence            5939
dias_contrato_restante    5939
dtype: int64

Cardinalidad de columnas categóricas / booleanas:
  Nation                      134 categorías
  Pos                          10 categorías
  Squad                       130 categorías
  Comp                          5 categorías
  Season                        5 categorías
  posicion_tm                  13 categorías
  club                        130 categorías
  club_coincide                 2 categorías
  flag_sin_tiros                2 categorías
  flag_sin_titularidades        2 categorías
  flag_sin_suplencias           2 categorías
  flag_cruce_verificado         2 categorías
  flag_contrato_conocido        2 categorías


## 2. Corrección de cruce y eliminación de identificadores/auditoría

### 2.1 Filas con cruce FBref↔Transfermarkt ambiguo

Al revisar los residuos de un Random Forest en `04-modelado.ipynb`, las predicciones para "Rodri" eran inconsistentes entre filas que decían tener el mismo `valor_eur`. La causa está en `01-unificacion.ipynb`: el cruce FBref↔Transfermarkt se hace por nombre normalizado (con año de nacimiento como *fallback*), y el merge final (`fb.merge(tmu, on=['player_id','saison_id'], how='left', validate='many_to_one')`) permite que **más de una fila de FBref** —jugadores distintos que comparten nombre normalizado— quede pegada al mismo `player_id` (y por lo tanto al mismo `valor_eur`) de Transfermarkt en la misma temporada.

Esto afecta a **735 grupos `(player_id, saison_id)`** (1.473 filas, ~10.6% del dataset) con más de un `Squad` distinto para el mismo `player_id` en la misma temporada — es decir, estadísticas de un jugador equivocado pegadas al valor de mercado de otro. La columna `club_coincide` (¿el `Squad` de FBref coincide con el `club` de Transfermarkt de esa temporada?), calculada en el cruce, permite distinguir la fila correcta:

- **660 grupos**: exactamente una fila tiene `club_coincide=True` → se conserva esa fila (la correcta) y se descartan las demás del grupo.
- **75 grupos**: ninguna fila coincide con el club de Transfermarkt → no hay forma confiable de saber cuál (si alguna) es la correcta, se descarta el grupo completo.

Este tratamiento se aplica **antes** de descartar `club_coincide` como columna de auditoría (sección 2.2), porque hace falta para esta limpieza.

### 2.2 Identificadores, columnas de auditoría y fuga de información

| Columna | Tratamiento | Motivo |
|---|---|---|
| `Rk` | Eliminar | Índice de scraping sin significado, no es un atributo del jugador. |
| `nom` | Eliminar | Nombre normalizado en minúsculas, usado solo como clave de cruce FBref↔Transfermarkt (redundante con `Player`). |
| `pid_fb` | Eliminar | Duplica `player_id` en el 99.8% de los casos; es un campo de auditoría del cruce, no un atributo. |
| `anio_nac` | Eliminar | Duplica `Born`/`nacimiento`, y ambos a su vez son redundantes con `Age`. |
| `Born` | Eliminar | Redundante con `Age` (año de nacimiento vs. edad calculada); se conserva `Age`. |
| `nacimiento` | Eliminar | La información relevante (edad) ya queda resumida en `Age`; el día/mes exacto de nacimiento no aporta valor de mercado adicional. |
| `club` | Eliminar | Según el EDA, es una *instantánea* del club actual en Transfermarkt, no un histórico por temporada (puede no coincidir con el club real en temporadas antiguas). Se usa `Squad`, que sí es específico de cada temporada de FBref. |
| `Season` | Eliminar | Redundante con `saison_id` (mismo valor pero en formato texto "2021-2022" vs. entero 2021); se conserva la versión numérica. |
| `flag_cruce_verificado` | Eliminar | Flag de auditoría del cruce de fuentes (si `pid_fb` existía o no) — describe la calidad del dato, no una característica del jugador que explique su valor de mercado. |
| `club_coincide` | Eliminar (ya usada en 2.1) | Igual que el anterior: flag de calidad de datos, no un atributo del jugador — pero antes de eliminarla se usó para resolver los cruces ambiguos de la sección 2.1. |
| `n_metodos_atipico` | **Eliminar (fuga de información)** | En el EDA se calculó ejecutando IsolationForest / LOF / EllipticEnvelope / Mahalanobis usando **`valor_eur` como una de las variables de entrada**. Usarlo como *feature* filtraría directamente información del objetivo al modelo. |

`player_id` y `Player` se conservan, pero **solo como identificadores** (no como *features* de entrada al modelo): permiten trazabilidad y, como señala el EDA, dividir train/test por `player_id` para evitar que el mismo jugador aparezca en ambos conjuntos.

In [6]:
key_cols = ["player_id", "saison_id"]
n_squads = df.groupby(key_cols)["Squad"].transform("nunique")
ambiguos = n_squads > 1
n_true_grupo = df.groupby(key_cols)["club_coincide"].transform("sum")

n_grupos_ambiguos = df.loc[ambiguos, key_cols].drop_duplicates().shape[0]
n_grupos_resueltos = df.loc[ambiguos & (n_true_grupo == 1), key_cols].drop_duplicates().shape[0]
n_grupos_sin_confianza = df.loc[ambiguos & (n_true_grupo == 0), key_cols].drop_duplicates().shape[0]

print(f"Grupos (player_id, saison_id) con Squad ambiguo: {n_grupos_ambiguos}")
print(f"  - resueltos via club_coincide (queda 1 fila correcta): {n_grupos_resueltos}")
print(f"  - sin ninguna fila confiable, se descarta el grupo completo: {n_grupos_sin_confianza}")

descartar = (
    (ambiguos & (n_true_grupo == 1) & (~df["club_coincide"]))
    | (ambiguos & (n_true_grupo == 0))
)

print(f"\nFilas descartadas: {descartar.sum()} de {df.shape[0]}")
df = df[~descartar].reset_index(drop=True)
print(f"Shape actual: {df.shape}")

Grupos (player_id, saison_id) con Squad ambiguo: 735
  - resueltos via club_coincide (queda 1 fila correcta): 660
  - sin ninguna fila confiable, se descarta el grupo completo: 75

Filas descartadas: 813 de 13835
Shape actual: (13022, 95)


In [7]:
cols_identificadores_auditoria = [
    "Rk", "nom", "pid_fb", "anio_nac", "Born", "nacimiento", "club", "Season",
    "flag_cruce_verificado", "club_coincide",
]
cols_leakage = ["n_metodos_atipico"]

df = df.drop(columns=cols_identificadores_auditoria + cols_leakage)
print(f"Eliminadas {len(cols_identificadores_auditoria)} columnas de identificadores/auditoría "
      f"y {len(cols_leakage)} columna con fuga de información.")
print(f"Shape actual: {df.shape}")

Eliminadas 10 columnas de identificadores/auditoría y 1 columna con fuga de información.
Shape actual: (13022, 84)


## 3. Duplicados exactos y variables derivadas (combinación lineal)

FBref repite algunas métricas en varias tablas de su web (todas fueron scrapeadas y quedaron en el dataset con nombres distintos). Verificamos que son **idénticas fila a fila** antes de eliminarlas, y lo mismo para columnas que son resultado de sumar/restar otras dos que también se conservan (colinealidad perfecta, no solo alta correlación):

| Columna a eliminar | Motivo (idéntica o derivada de...) |
|---|---|
| `Standard_Gls` | = `Performance_Gls` (100% igual) |
| `Standard_PK` | = `Performance_PK` (100% igual) |
| `Standard_PKatt` | = `Performance_PKatt` (100% igual) |
| `90s` | = `Playing Time_90s` (100% igual) |
| `Starts_Starts` | = `Playing Time_Starts` (100% igual) |
| `Performance_G+A` | = `Performance_Gls` + `Performance_Ast` |
| `Performance_G-PK` | = `Performance_Gls` − `Performance_PK` |
| `Per 90 Minutes_G+A` | = `Per 90 Minutes_Gls` + `Per 90 Minutes_Ast` |
| `Per 90 Minutes_G-PK` | = versión por-90 de `Performance_Gls` − `Performance_PK` (además llegó como texto en el scraping original) |
| `Per 90 Minutes_G+A-PK` | = versión por-90 de `Performance_Gls` + `Performance_Ast` − `Performance_PK` |
| `Team Success_+/-` | = `Team Success_onG` − `Team Success_onGA` |
| `Team Success_+/-90` | = `Team Success_+/-` normalizado por `90s` |

En todos los casos se conserva **una sola** de las columnas equivalentes (la de nombre más simple/estándar) y se descartan las derivadas, ya que no aportan información nueva y solo inflarían la multicolinealidad.

In [8]:
# Verificación de igualdad exacta antes de eliminar
dup_pairs = [
    ("Standard_Gls", "Performance_Gls"),
    ("Standard_PK", "Performance_PK"),
    ("Standard_PKatt", "Performance_PKatt"),
    ("90s", "Playing Time_90s"),
    ("Starts_Starts", "Playing Time_Starts"),
]
for a, b in dup_pairs:
    print(f"{a:18s} == {b:22s} -> {(df[a] == df[b]).mean():.3f} de las filas iguales")

cols_duplicadas = [a for a, _ in dup_pairs]
cols_derivadas = [
    "Performance_G+A", "Performance_G-PK",
    "Per 90 Minutes_G+A", "Per 90 Minutes_G-PK", "Per 90 Minutes_G+A-PK",
    "Team Success_+/-", "Team Success_+/-90",
]

df = df.drop(columns=cols_duplicadas + cols_derivadas)
print(f"\nEliminadas {len(cols_duplicadas)} columnas duplicadas y {len(cols_derivadas)} derivadas.")
print(f"Shape actual: {df.shape}")

Standard_Gls       == Performance_Gls        -> 1.000 de las filas iguales
Standard_PK        == Performance_PK         -> 1.000 de las filas iguales
Standard_PKatt     == Performance_PKatt      -> 1.000 de las filas iguales
90s                == Playing Time_90s       -> 1.000 de las filas iguales
Starts_Starts      == Playing Time_Starts    -> 1.000 de las filas iguales

Eliminadas 5 columnas duplicadas y 7 derivadas.
Shape actual: (13022, 72)


## 4. Redundancia en el "volumen de participación"

`Playing Time_MP` (partidos jugados), `Playing Time_Starts` (titularidades), `Playing Time_90s` y `Starts_Compl` (partidos completos) están todas correlacionadas por encima de 0.88 con `Playing Time_Min` (minutos totales, `Playing Time_90s` incluso al 1.00 por construcción): son formas casi intercambiables de medir cuánto jugó un futbolista en la temporada. Mantener las cinco no añade señal nueva y sí multicolinealidad.

**Tratamiento:** se conserva únicamente `Playing Time_Min` (a través de su versión ya escalada `Playing Time_Min_rs`, ver sección 5) como proxy del volumen de participación, y se eliminan las demás.

In [9]:
for c in ["Playing Time_MP", "Playing Time_Starts", "Playing Time_90s", "Starts_Compl"]:
    print(f"corr({c}, Playing Time_Min) = {df[c].corr(df['Playing Time_Min']):.3f}")

cols_volumen_redundante = ["Playing Time_MP", "Playing Time_Starts", "Playing Time_90s", "Starts_Compl"]
df = df.drop(columns=cols_volumen_redundante)
print(f"\nShape actual: {df.shape}")

corr(Playing Time_MP, Playing Time_Min) = 0.915
corr(Playing Time_Starts, Playing Time_Min) = 0.994
corr(Playing Time_90s, Playing Time_Min) = 1.000
corr(Starts_Compl, Playing Time_Min) = 0.879

Shape actual: (13022, 68)


## 5. Relativizar variables ofensivas y defensivas por posición

`Performance_Gls`, `Performance_Ast`, `Standard_Sh` (producción ofensiva) y `Performance_TklW`, `Performance_Int` (involucramiento defensivo) significan cosas muy distintas según la posición: 5 goles es excepcional para un central pero mediocre para un delantero, y el argumento espejo aplica a entradas/intercepciones (normal para un mediocentro defensivo, raro para un delantero. Se reemplaza el escalado global de estas 5 variables por un `RobustScaler` ajustado **dentro de cada `posicion_tm`**, para que cada valor se compare contra jugadores de su misma posición y no contra todo el dataset. Tiene que calcularse acá, antes de que la sección 8 convierta `posicion_tm` en dummies y la descarte.

**Nota técnica:** para porteros estas 5 estadísticas son prácticamente siempre 0 (mediana e IQR = 0 dentro del grupo). `RobustScaler` ya maneja ese caso de forma segura (deja `scale_ = 1` en vez de dividir por 0), así que no hace falta un tratamiento especial aparte -- simplemente quedan en 0 tras el escalado, correcto para un grupo sin varianza.

In [10]:
from sklearn.preprocessing import RobustScaler

STATS_POR_POSICION = ["Performance_Gls", "Performance_Ast", "Standard_Sh", "Performance_TklW", "Performance_Int"]

partes = []
for pos, sub in df.groupby("posicion_tm", observed=True):
    rs = RobustScaler()
    escalado = rs.fit_transform(sub[STATS_POR_POSICION])
    partes.append(pd.DataFrame(escalado, columns=[f"{c}_rs" for c in STATS_POR_POSICION], index=sub.index))

df_pos_rs = pd.concat(partes).sort_index()
df[df_pos_rs.columns] = df_pos_rs  # sobreescribe Gls/Ast/Sh (antes globales) y crea TklW/Int (nuevas)

# Cada posición queda centrada en su propia mediana (~0), a diferencia del escalado global
print("Mediana de Performance_Gls_rs por posición (debería rondar 0 en todas):")
print(df.groupby("posicion_tm", observed=True)["Performance_Gls_rs"].median().round(2))

Mediana de Performance_Gls_rs por posición (debería rondar 0 en todas):
posicion_tm
Attacking Midfield    0.0
Central Midfield      0.0
Centre-Back           0.0
Centre-Forward        0.0
Defensive Midfield    0.0
Goalkeeper            0.0
Left Midfield         0.0
Left Winger           0.0
Left-Back             0.0
Right Midfield        0.0
Right Winger          0.0
Right-Back            0.0
Second Striker        0.0
Name: Performance_Gls_rs, dtype: float64


In [11]:
# Performance_Gls/Ast/Standard_Sh (crudos) los elimina la sección 6 más abajo (ya
# reemplazados por su _rs, ahora por posición). TklW/Int no tenían tratamiento previo:
df = df.drop(columns=["Performance_TklW", "Performance_Int"])
print(f"Shape actual: {df.shape}")

Shape actual: (13022, 68)


## 6. Consolidar variables "en bruto" con su versión ya tratada en el EDA

El EDA (`02-eda.ipynb`) ya generó, para las variables más sesgadas, una versión transformada:

- **`_rs`** (RobustScaler: `(x − mediana) / IQR`) para `Age`, `altura_m`, `Playing Time_Min`, `Performance_Gls`, `Performance_Ast`, `Standard_Sh`, `Performance_CrdY`, `Team Success_PPM`.
- **`log1p_`** para variables de conteo muy asimétricas: `Performance_Gls`, `Performance_Ast`, `Standard_Sh`, `Standard_SoT`, `Performance_Crs`, `Performance_Fls`.

Mantener a la vez la columna original, su `log1p_` y su `_rs` sería usar **la misma señal tres veces** en distintas escalas. Se define una única representación final por variable:

| Variable base | Representación final elegida | Se descartan |
|---|---|---|
| `Age` | `Age_rs` | `Age` |
| `altura_m` | `altura_m_rs` | `altura_m` |
| `Playing Time_Min` | `Playing Time_Min_rs` | `Playing Time_Min` |
| `Performance_CrdY` | `Performance_CrdY_rs` | `Performance_CrdY` |
| `Team Success_PPM` | `Team Success_PPM_rs` | `Team Success_PPM` |
| `Performance_Gls` | `Performance_Gls_rs` | `Performance_Gls`, `log1p_Performance_Gls` |
| `Performance_Ast` | `Performance_Ast_rs` | `Performance_Ast`, `log1p_Performance_Ast` |
| `Standard_Sh` | `Standard_Sh_rs` | `Standard_Sh`, `log1p_Standard_Sh` |
| `Standard_SoT` | `log1p_Standard_SoT` | `Standard_SoT` |
| `Performance_Crs` | `log1p_Performance_Crs` | `Performance_Crs` |
| `Performance_Fls` | `log1p_Performance_Fls` | `Performance_Fls` |

Criterio: para las variables con `_rs` disponible se usa el escalado robusto (mediana/IQR), que ya vimos en el EDA que preserva mejor la forma de la distribución sin comprimir la señal de los jugadores más usados/productivos. Para las que solo cuentan con `log1p_` se usa esa versión, que corrige directamente la asimetría de la variable de conteo.

In [12]:
cols_reemplazadas_por_rs = [
    "Age", "altura_m", "Playing Time_Min", "Performance_CrdY", "Team Success_PPM",
]
cols_reemplazadas_por_rs_y_log1p = ["Performance_Gls", "Performance_Ast", "Standard_Sh"]
cols_log1p_duplicado = [f"log1p_{c}" for c in cols_reemplazadas_por_rs_y_log1p]
cols_reemplazadas_por_log1p = ["Standard_SoT", "Performance_Crs", "Performance_Fls"]

df = df.drop(columns=(
    cols_reemplazadas_por_rs
    + cols_reemplazadas_por_rs_y_log1p
    + cols_log1p_duplicado
    + cols_reemplazadas_por_log1p
))
print(f"Shape actual: {df.shape}")

Shape actual: (13022, 54)


## 7. `Nation`: alta cardinalidad, se descarta

`Nation` tiene 134 categorías, muchas con muy pocas observaciones. Convertirla en dummies añadiría más de 130 columnas casi todas dispersas (mayoría de ceros), con alto riesgo de sobreajuste y sin una justificación futbolística clara de que la nacionalidad, controlando por liga/equipo/rendimiento, explique el valor de mercado.

**Tratamiento:** se elimina. La señal relevante de "mercado/prestigio" que sí queremos capturar queda representada por variables de más bajo costo dimensional: la competición (`Comp`), el equipo (`Squad`) y los indicadores `Ganador_Champions` / `Ganador_Liga` ya construidos.

In [13]:
print(f"Nation: {df['Nation'].nunique()} categorías")
df = df.drop(columns=["Nation"])
print(f"Shape actual: {df.shape}")

Nation: 134 categorías
Shape actual: (13022, 53)


## 8. Variables categóricas: posición, competición y equipo

Se evalúan tres categóricas: `posicion_tm`, `Comp` y `Squad`. Las dos primeras tienen cardinalidad baja (13 y 5 categorías) y se codifican con one-hot (`pd.get_dummies`, `drop_first=True` para evitar la *dummy variable trap* en modelos lineales/regularizados; en árboles no supone pérdida de información). `Squad`, con 130 categorías, recibe un tratamiento distinto (ver 8.3).

### 8.1 Posición (`posicion_tm`)
`Pos` (FBref) y `posicion_tm` (Transfermarkt) son consistentes entre sí (ver heatmap del EDA), pero `posicion_tm` es más granular (13 categorías puras: `Centre-Back`, `Left Winger`, `Defensive Midfield`, …) mientras que `Pos` mezcla combinaciones (`DF,MF`, `MF,FW`) para jugadores multiposición. Se usa **`posicion_tm`** como fuente de la dummy de posición y se elimina `Pos` para no duplicar la misma información con una codificación más ambigua.

### 7.2 Competición (`Comp`)
Solo 5 categorías (las 5 grandes ligas). El EDA mostró diferencias claras en la mediana de `log_valor_eur` por liga (Premier League y La Liga por encima del resto), por lo que aporta señal directa de "nivel de mercado" de cada liga. Se codifica completa en dummies.

### 7.3 Equipo (`Squad`): por qué no one-hot

`Squad` tiene 130 categorías. Un one-hot completo generaría ~129 columnas casi siempre dispersas, muchas con apenas 1-2 observaciones (equipos que solo pasaron una temporada por las 5 grandes ligas). Agrupar en un top-N + `"Otro_equipo"` (como se hacía en una versión anterior de este notebook) reduce la dimensionalidad, pero colapsa la mayoría de los equipos en una sola categoría, perdiendo toda la señal individual de clubes medianos (p. ej. Betis, Villarreal, Eintracht Frankfurt) que sí tienen un efecto propio y distinto sobre el valor de mercado.

Opciones evaluadas para reemplazar el one-hot:

| Método | Cómo funciona | Ventajas | Problema para este caso |
|---|---|---|---|
| **Label Encoding** | Cada equipo → un entero arbitrario (0, 1, 2, …) | Trivial, una sola columna | Introduce una relación de orden falsa entre equipos que no existe (`Zaragoza=45` no es "más" que `Betis=12`); modelos lineales la tomarían literalmente, y hasta en árboles el orden arbitrario del encoder puede sesgar los splits |
| **Frequency / Count Encoding** | Cada equipo → nº de filas jugador-temporada que tiene en el dataset | Una sola columna, no usa el target (cero riesgo de fuga), robusta | Mide *tamaño de plantilla / permanencia en el dataset*, no *prestigio o valor de mercado*: dos equipos con frecuencia similar pueden tener valores de mercado muy distintos |
| **Target Encoding (mean encoding suavizado)** | Cada equipo → media de `log_valor_eur` para ese equipo, regularizada hacia la media global según su frecuencia (así los equipos con pocas observaciones no quedan con una media ruidosa) | Captura directamente la señal que motivaba el one-hot (los clubes "grandes" inflan el valor de sus jugadores), conserva las 130 categorías en una sola columna | Usa el target: si se calcula sobre todo el dataset **antes** de separar train/test, filtra información del test hacia las features de train |

**Decisión:** Target Encoding es el que mejor sustituye la señal que buscaba capturar el one-hot, sin la explosión dimensional del one-hot ni la pérdida de información del agrupamiento top-N. Para no introducir fuga de información, **no se calcula en este notebook**: aquí `Squad` se deja como columna categórica sin codificar, y el target encoding se ajusta en `04-modelado.ipynb` **solo con los datos de train**, después del split por `player_id` (con `sklearn.preprocessing.TargetEncoder`, que además aplica *cross-fitting* interno para no sobreajustar dentro del propio train). Es el mismo criterio de rigor que ya se aplicó en la sección 2 con `n_metodos_atipico`.

In [14]:
# 7.1 Posición: dummies desde posicion_tm, se elimina Pos (redundante, menos granular)
df = df.drop(columns=["Pos"])
dummies_pos = pd.get_dummies(df["posicion_tm"], prefix="pos", drop_first=True)

# 7.2 Competición: dummies desde Comp
dummies_comp = pd.get_dummies(df["Comp"], prefix="liga", drop_first=True)

print(f"Dummies de posición: {dummies_pos.shape[1]}")
print(f"Dummies de competición: {dummies_comp.shape[1]}")

df = pd.concat([df, dummies_pos, dummies_comp], axis=1)
df = df.drop(columns=["posicion_tm", "Comp"])

# 7.3 Equipo: se deja `Squad` como categórica sin codificar (130 categorías).
# El target encoding se ajusta en 04-modelado.ipynb solo con el train, tras el split.
df["Squad"] = df["Squad"].astype(str)
print(f"\nSquad se conserva sin codificar: {df['Squad'].nunique()} categorías (pendiente de target encoding en 04-modelado.ipynb)")

print(f"\nShape actual: {df.shape}")

Dummies de posición: 12
Dummies de competición: 4

Squad se conserva sin codificar: 130 categorías (pendiente de target encoding en 04-modelado.ipynb)

Shape actual: (13022, 66)


## 9. Vencimiento de contrato: representación final

`contrato_vence` es una fecha cruda -- no es una *feature* utilizable tal cual (ni los modelos lineales ni los de árbol de este proyecto aceptan `datetime`). La señal numérica ya vive en `dias_contrato_restante` (derivada en `02-eda.ipynb`), así que se elimina la fecha cruda y **solo falta decidir qué hacer con su 43% de nulos**, ya que `LinearRegression` y `RandomForestRegressor` no toleran `NaN` (XGBoost sí, pero hay que dejar una sola matriz de *features* válida para los cinco modelos de `04-modelado.ipynb`).

**Tratamiento:** se imputa con la mediana global (elección neutra: no asume que el jugador con contrato desconocido tiene *más* o *menos* tiempo restante que el resto) y se conserva `flag_contrato_conocido` (ya creada) para que el modelo pueda aprender si el hecho de no tener el dato es, en sí mismo, informativo.

In [15]:
mediana_dias = df["dias_contrato_restante"].median()
df["dias_contrato_restante"] = df["dias_contrato_restante"].fillna(mediana_dias)
df = df.drop(columns=["contrato_vence"])

print(f"Mediana usada para imputar: {mediana_dias:.0f} días")
print(f"Nulos restantes en dias_contrato_restante: {df['dias_contrato_restante'].isna().sum()}")
print(f"Shape actual: {df.shape}")

Mediana usada para imputar: 1063 días
Nulos restantes en dias_contrato_restante: 0
Shape actual: (13022, 65)


## 10. Tipos finales y guardado

Se convierten las columnas booleanas (flags e indicadores dummy) a enteros (`0`/`1`) para tener una matriz mayormente numérica, y se guarda el resultado. `Squad` queda como única columna categórica sin codificar (ver sección 8.3): no es un *feature* utilizable tal cual, pero se conserva para que `04-modelado.ipynb` calcule su target encoding correctamente, ajustándolo solo sobre el train tras el split.

**Variables que quedan en `df` y su rol:**

| Rol | Columnas |
|---|---|
| Identificadores (no usar como *feature*) | `Player`, `player_id`, `saison_id` |
| Objetivo | `valor_eur`, `log_valor_eur` (usar `log_valor_eur` como variable a predecir; `valor_eur` se conserva para interpretar resultados en euros) |
| Categórica pendiente de codificar | `Squad` (130 categorías) — target encoding a ajustar en `04-modelado.ipynb` sobre el train, después del split, para evitar fuga de información del test |
| Numéricas ya tratadas (`_rs` / `log1p_`) | `Age_rs`, `altura_m_rs`, `Playing Time_Min_rs`, `Performance_Gls_rs`, `Performance_Ast_rs`, `Standard_Sh_rs`, `Performance_CrdY_rs`, `Team Success_PPM_rs`, `log1p_Standard_SoT`, `log1p_Performance_Crs`, `log1p_Performance_Fls` |
| Numéricas de rendimiento sin redundancia detectada | `Performance_PK`, `Performance_PKatt`, `Performance_CrdR`, `Performance_2CrdY`, `Performance_Fld`, `Performance_Off`, `Performance_Int`, `Performance_TklW`, `Performance_OG`, `Per 90 Minutes_Gls`, `Per 90 Minutes_Ast`, `Standard_SoT%`, `Standard_Sh/90`, `Standard_SoT/90`, `Standard_G/Sh`, `Standard_G/SoT`, `Playing Time_Mn/MP`, `Playing Time_Min%`, `Starts_Mn/Start`, `Subs_Subs`, `Subs_Mn/Sub`, `Subs_unSub`, `Team Success_onG`, `Team Success_onGA`, `Team Success_On-Off` |
| Flags indicadoras (del EDA) | `flag_sin_tiros`, `flag_sin_titularidades`, `flag_sin_suplencias` |
| Indicadores de éxito de equipo (creados en este notebook) | `Ganador_Champions`, `Ganador_Liga` |
| Categóricas → dummies | `pos_*` (posición), `liga_*` (competición) |

Nota: las variables de rendimiento sin redundancia siguen mostrando cierta correlación cruzada entre sí (p. ej. cluster ofensivo goles/tiros/asistencias vs. cluster defensivo tarjetas/faltas/intercepciones, ya identificado en la red de correlación del EDA). No se eliminó ninguna adicional porque no son duplicados exactos ni combinaciones lineales — cada una describe un aspecto de juego distinto — pero se recomienda, en la etapa de modelado, usar regularización (Ridge/Lasso) o revisar importancia de variables (modelos de árboles / SHAP) para depurarlas si fuera necesario.

In [16]:
bool_cols = df.select_dtypes(include="bool").columns
df[bool_cols] = df[bool_cols].astype(int)

print(f"Shape final: {df.shape}")
print(f"Columnas booleanas convertidas a int: {len(bool_cols)}")
print()
print("Nulos restantes:", df.isna().sum().sum())
print()
df.dtypes.value_counts()

Shape final: (13022, 65)
Columnas booleanas convertidas a int: 20

Nulos restantes: 0



int64      35
float64    27
str         3
Name: count, dtype: int64

In [17]:
from pathlib import Path

OUT_PATH = Path("../data/processed/fbref_tm_features.parquet")
df.to_parquet(OUT_PATH, index=False)
print(f"Dataset de variables para modelado guardado en: {OUT_PATH.resolve()}")
print(f"Shape: {df.shape}")

Dataset de variables para modelado guardado en: /home/diego/diplomado/data/processed/fbref_tm_features.parquet
Shape: (13022, 65)
